### ISSUE TITLE
CF doesn't pick the CMD Directive in the Dockerfile

### ISSUE THREAD

**Issue by rjain-pivotal** [cloudfoundry-incubator/diego-design-notes#27]
I was running a nodejs image which has the following CMD directive
```javascript
FROM centos:centos6
# Enable EPEL for Node.js
RUN     rpm -Uvh http://download.fedoraproject.org/pub/epel/6/i386/epel-release-6-8.noarch.rpm
# Install Node.js and npm
RUN     yum install -y npm
COPY . /src
RUN cd /src; npm install
EXPOSE 8080
CMD ["node", "/src/index.js"]
```

* CF Push without -c won't work
`cf push --docker-image rjain15/centos-node-hello "`

* This works: 
`cf push --docker-image rjain15/centos-node-hello  -c "node /src/index.js"`

**Comment by ematpl** [comment]
Hi, @rjain-pivotal,

Thanks for the report. Just to confirm, are you seeing failures like the following in your `cf logs` output, or is this a different error?

```
2015-10-28T10:42:53.21-0700 [STG/0]      OUT Creating container
2015-10-28T10:42:53.62-0700 [STG/0]      OUT Successfully created container
2015-10-28T10:42:53.79-0700 [STG/0]      OUT Staging...
2015-10-28T10:42:53.81-0700 [STG/0]      OUT Staging process started ...
2015-10-28T10:42:54.56-0700 [STG/0]      OUT Staging process finished
2015-10-28T10:42:54.56-0700 [STG/0]      OUT Exit status 0
2015-10-28T10:42:54.56-0700 [STG/0]      OUT Staging Complete
2015-10-28T10:42:55.90-0700 [CELL/0]     OUT Creating container
2015-10-28T10:43:20.19-0700 [CELL/0]     OUT Successfully created container
2015-10-28T10:43:20.38-0700 [APP/0]      OUT Exit status 0
2015-10-28T10:43:20.39-0700 [CELL/0]     OUT Exit status 2
2015-10-28T10:43:20.45-0700 [API/0]      OUT App instance exited with guid 92c81c2a-1e28-428c-aef3-7333f2befcb1 payload: {"instance"=>"76054135-3c1d-49cb-4be5-f8dea83a8e30", "index"=>0, "reason"=>"CRASHED", "exit_description"=>"2 error(s) occurred:\n\n* 2 error(s) occurred:\n\n* Codependent step exited\n* cancelled\n* cancelled", "crash_count"=>1, "crash_timestamp"=>1446054200430043931, "version"=>"b815aef3-3c5c-47e9-a39d-6cf38034cf31"}
2015-10-28T10:43:20.54-0700 [CELL/0]     OUT Creating container
2015-10-28T10:44:28.47-0700 [CELL/0]     OUT Successfully created container
2015-10-28T10:44:28.72-0700 [APP/0]      OUT Exit status 0
2015-10-28T10:44:28.73-0700 [CELL/0]     OUT Exit status 2
2015-10-28T10:44:28.80-0700 [API/0]      OUT App instance exited with guid 92c81c2a-1e28-428c-aef3-7333f2befcb1 payload: {"instance"=>"ca2b5d36-7626-458b-4a09-a9237fcdb98a", "index"=>0, "reason"=>"CRASHED", "exit_description"=>"2 error(s) occurred:\n\n* 2 error(s) occurred:\n\n* Codependent step exited\n* cancelled\n* cancelled", "crash_count"=>2, "crash_timestamp"=>1446054268772665450, "version"=>"b815aef3-3c5c-47e9-a39d-6cf38034cf31"}
```

That log output is from my pushing your image to one of the Diego team's test environments. From inspecting the Diego LRP created for that CF app, we are correctly picking up the CMD and ENTRYPOINT directives from the Docker image metadata, but for some reason when the launcher from the docker-app-lifecycle execs the command, it exits 0 (that's the APP line with "Exit status 0" above) instead of running the desired process. We've seen this error with some Java-based containers as well and confirmed that the workaround of specifying the start command explicitly does work, but obviously we'd like to autodetect it from the Docker image metadata. I've created https://www.pivotaltracker.com/story/show/106873182 for the Diego team to investigate and resolve the issue. Do you mind if I copy your Docker image into a separate account to use as another reproduction case on that story?

Thanks,
Eric, CF Runtime Diego PM

**Comment by datianshi** [comment]
https://hub.docker.com/r/datianshi/cloud9-docker/

@ematpl I got the similar issue. you can use mine for trouble shooting

**Comment by ematpl** [comment]
Thanks, @datianshi! Do you also have the Dockerfile used to build this image? If we use it as a reproduction case, that would help us validate that we're running the intended ENTRYPOINT + CMD directives as the start command.

Best,
Eric

**Comment by rjain-pivotal** [comment]
Sorry guys, seems the message got lost in all the other mailing list crap.
I will attach the Dockerfile if I can get hold of it, or try to recreate
it.

On Tue, Nov 3, 2015 at 7:57 PM, Eric Malm <notifications@github.com> wrote:

> Thanks, @datianshi <https://github.com/datianshi>! Do you also have the
> Dockerfile used to build this image? If we use it as a reproduction case,
> that would help us validate that we're running the intended ENTRYPOINT +
> CMD directives as the start command.
>
> Best,
> Eric
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cloudfoundry-incubator/diego-design-notes/issues/27#issuecomment-153537244>
> .
>



-- 
Rajesh Jain
Sales Engineering, Pivotal
e:rajesh.jain@pivotal.io <rajesh.jain@pivotal.io> | m: 2487900082 | t:
@rjain15

**Comment by datianshi** [comment]
https://github.com/datianshi/cloud9-docker/blob/master/Dockerfile

cf push desktop -o datianshi/cloud9-docker -c "node /cloud9/server.js -l 0.0.0.0 -p 8080 -w /workspace -a test:test"

This seems work

**Comment by ematpl** [comment]
Thanks, @datianshi. I see that the start command you have in the `cf push` command above differs from the CMD directive in the Dockerfile. Does the app also start up if you push it using `-c "node /cloud9/server.js --listen 0.0.0.0 --port 8080 -w /workspace"` instead?

Best,
Eric

**Comment by datianshi** [comment]
Yes it works in both -c and docker run 
but not docker push directly 

Sorry about the confusion on acronym of the command

**Comment by cwlbraa** [comment]
Hi. We think we addressed this problem with https://github.com/cloudfoundry-incubator/docker_app_lifecycle/commit/4f8aa35a149a75596a32cb8469eb66b106d29cf3.  Please let us know if you're still having issues.

Thanks.

**Comment by ematpl** [comment]
To be clear, this work is still in flight, and we're expecting this to be fixed fully in Diego final release 0.1441.0.
